File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-2.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {    
    public bool IsSubtree(TreeNode root, TreeNode subRoot) {
        if(subRoot==null) return true;
        if(root==null) return false;
        if(root.val==subRoot.val && IsSameTree(root.left,subRoot.left) && IsSameTree(root.right, subRoot.right)){
            return true;
        }
        else{
            return IsSubtree(root.left, subRoot)||IsSubtree(root.right, subRoot);
        }
    }

    public bool IsSameTree(TreeNode t1, TreeNode t2){
        if(t1==null && t2==null) return true;
        if(t1!=null && t2==null) return false;
        if(t1==null && t2!=null) return false;
        return (t1.val==t2.val && IsSameTree(t1.left,t2.left) && IsSameTree(t1.right,t2.right));
    }
}

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int goodNodes(TreeNode root) {
        int count = 0;
        Queue<Pair<TreeNode, Integer>> q = new LinkedList<>();
        q.add(new Pair<>(root, Integer.MIN_VALUE));

        while(!q.isEmpty()){
            int len = q.size();
            for(int i=0;i<len;i++){
                Pair<TreeNode, Integer> ele = q.poll();
                TreeNode node = ele.getKey();
                int maxSeen = ele.getValue();
                if(node.val>=maxSeen) count++;
                maxSeen = Math.max(maxSeen, node.val);
                if(node.left!=null) q.add(new Pair<>(node.left, maxSeen));
                if(node.right!=null) q.add(new Pair<>(node.right, maxSeen));
            }
        }
        return count;
    }
}
